[17299] Upload attachments in create form split view/edit view

floriank opened the pull request:
This is the first iteration of the work package attachment creation after creating it via the work package details pane in the split view.

Expect this to be heavily buggy. I only created this PR so that we get a teatro stage and discuss ideas (also, I get feedback from @houndci). All feedback and kitty pictures welcome, but keep in mind that I am striving for a more generic solution first with added polish (i.e. nail polish) later.

Should deal with most of the stuff in https://community.openproject.org/work_packages/17299.

This PR also introduces a very basic notification system based on `$rootScope.$broadcast`.

---
## TODO
- [x] Styling the file list
- [x] lock the user from leaving the page while uploads are in progress
- [x] Styling in general
- [x] Validation for file size(s) ( see #3244)
- [x] Labels should of course be `I18n` translations, this is not yet the case for all labels in the feature

Commit:
introduce ngFileupload

Commit:
provide a simple interface to files

Also, display a list. this also contains a simulation of how file descriptions
could be added.

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
add a basic upload service

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
add notification service with a centralized directive

this will enable notifications to be sent from anyone using the service,
displayed by the directive. It currently uses the $rootScope which is pretty
inefficient and should probably refactored to an pub/sub pattern style
implementation

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
remove dummy test code

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
emit a message when uploads are pending

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
probably fix create contract

Commit:
rebased against new notification styles

This will introduce some changes throughout the notifications. Upload
notifications can now group the uploads and the new notification system
can now receive removal commands.

In addition, I took the liberty of adding some minor animations

Signed-off-by: Florian Kraft <f.kraft@finn.de>

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://floriank-feature-17299-upload-attachments-in-create.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://floriank-feature-17299-upload-attachments-in-create.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

NobodysNightmare commented:
> Should we limit the maximum number of uploads?

Do you restrict the number of parallel uploads? I.e. will the client send 100 files simultaneously or do we have a limit to e.g. 5 files in parallel? This would be useful in general to avoid having 100 stalled uploads as opposed to 5 good uploads and 95 queued ones...

NobodysNightmare commented on frontend/app/ui_components/upload-progress-directive.js:
```diff
@@ -0,0 +1,58 @@
+        scope.value = Math.round(details.loaded / details.total * 100);
+      } else {
+        // dummy value if not computable
+        scope.value = 10;
```
Is there something like an indeterminate state for the progress control?

E.g. in .NET a progress bar can be determinate (indicating the progress based on value and max value) or indeterminate, showing a moving progress bar that can't be used to tell anything about the progress, but shows that "something is going on".

floriank replied:
You could show only the empty progress bar - which one may define styles for. For all our supported browsers this should not be a problem though, as they all provide computable file uploads.

NobodysNightmare commented on frontend/app/work_packages/services/work-package-attachments-service.js:
```diff
@@ -0,0 +1,69 @@
+  var currentUploads = [];
+
+  var upload = function(workPackage, files) {
+    var uploadPath = PathHelper.apiV3WorkPackageAttachmentsPath(workPackage.props.id);
```
My [latest PR for attachments](https://github.com/opf/openproject/pull/3244) adds the `addAttachment` link to a work package. That should allow you to write something like `workPackage.links.addAttachment.href`. I see few use of a client side path helper as a win ;-)

floriank replied:
Great! I will change this as soon as #3244 gets merged.

Commit:
use distinct event for uploading pending attachments

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
let the upload notification stick for a couple of seconds

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
add upload notification element

Commit:
basic upload for details view

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
wait with loading attachments until the uploads are done

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
move upload notification to notification component

Commit:
allow deletion of existing attachments

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
add to LSG different states of upload notification

Commit:
refactor upload notification HTML for LSG

Commit:
extend configuration service to use api endpoint

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
check for filesize

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
add styles for attachment list in WP tabs

Commit:
show file uploads in kilobytes

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
do not stop propagation

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
add a small directive to display an icon in the file list

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
add a download link for attachments

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
wrap the size definition into a helper method

the directive should not really care about whether an attachment is kB or MB
in size.

Signed-off-by: Florian Kraft <f.kraft@finn.de>

floriank commented:
This PR now, by accident (using the same component) also includes the handling of attachments in show mode covered in https://community.openproject.org/work_packages/20254

Commit:
fix style compatibility in all browsers supported

add missing class to attachment link

NobodysNightmare commented:
Just out of curiosity:

Whenever I click on the drop zone for files you are requesting the list-action for the WP attachments. Is that intentional? Does this have a purpose?

NobodysNightmare commented:
> ## TODO
> - [ ] Styling the file list

agreed! xD

(this is still very TODO ^^)

floriank commented:
Actually, the request to the attachments on click seems a bit bogus, let me investigate.

NobodysNightmare commented:
> This PR now, by accident (using the same component) also includes the handling of attachments in show mode

I only remember the talks and can't see the acceptance criteria. But didn't we say that the attachments **tab** should be removed as a consequence of showing the attachments in the overview?

You should probably do that in this PR as well, now that you already did the hard part ^^

NobodysNightmare commented:
> (this is still very TODO ^^)

Just pulled and now it looks awesome @vcostin ))

![niiiiice](https://cloud.githubusercontent.com/assets/453584/8823057/4405279a-306a-11e5-8d25-fbb7f45d358d.PNG)

Commit:
use correct mime type for jpg

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
do not ask the api for attachments when nothing changes

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
show an error notification instead of prolonging the file list

this will show files which cannot be uploaded in an error list in a notification
instead of using the file list.

Signed-off-by: Florian Kraft <f.kraft@finn.de>

floriank commented:
i actually prepared the list of mimetypes, but it does not strife to be complete. It's a bonus, so we can add mime types later.

Commit:
add a little space between the drop box and the potential files

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
get rid of the switch-case statement

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
clean up semicolon madness

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
correct the conversion service to actually work

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
typo and remove loading indicator for now

Signed-off-by: Florian Kraft <f.kraft@finn.de>

floriank commented:
In principle I am done with this initial feature spike.

Have a kitten:

![image](https://cloud.githubusercontent.com/assets/498241/8826559/9e1ae39c-3087-11e5-942f-134f61e4a275.png)

Commit:
get rid of ye olde attachments tab

Signed-off-by: Florian Kraft <f.kraft@finn.de>

NobodysNightmare commented:
Just noticed:

When clicking on an attachment my browser will always download it.
I remember that we talked about that. Just wanted to point out that:
- this is not how it worked before (in the legacy form)
- this is not how I like it to work w.r.t. pictures (I like pictures to be opened in my browser)

This comment is very specific and probably depends very much on my personal taste. Probably not the highest priority...

floriank commented:
Personaly, I would put off handling of downloads in this PR, as I:
1. Want pictures and pdfs to get a preview option in the fullscreen mode anyway
2. There is nothing specified really
3. The curent flow is consistent for all file types as of now.

NobodysNightmare commented:
> Personaly, I would put off handling of downloads in this PR

:+1:

ulferts commented on frontend/app/openproject-app.js:
```diff
@@ -44,7 +44,6 @@ require('angular-animate');
 
-// require('angular-i18n/angular-locale_en-us');
 if (I18n.locale === 'de') {
   require('angular-i18n/angular-locale_de-de');
```
Not that it is related to this PR but with all those languages now being part of the core, does this line still make sense or would it need to be generalized?

floriank replied:
the module is a bit misleading - `angular-i18n` provides localisation as well as internationalisation for different mesages and - most importantly - filters like `currency` and date.

I will investigate further, whether we really need this or if this can be considered cruft. However, this will not be done in the context for this PR :wink:

ulferts commented on frontend/app/templates/components/notification-box.html:
```diff
@@ -0,0 +1,35 @@
+            <i class="icon-arrow-right5-2"></i>
+          </a>
+          <span>
+            {{uploadCount}} of {{content.uploads.length}} files finished
```
I18n required.

Commit:
translate a missing label

[ci skip]

Signed-off-by: Florian Kraft <f.kraft@finn.de>

ulferts commented on frontend/app/ui_components/notification-box-directive.js:
```diff
@@ -0,0 +1,73 @@
+
+    scope.$on('upload.finished', function() {
+      scope.uploadCount += 1;
+    });
```
The whole controller seems to be highly upload specific. So either the controller needs to be renamed or it needs to be generalized.

floriank replied:
The directive _will_ manage more than just uploads (see the `notificationBox` directive template), so i think the generic name is appropiate.

ulferts commented on frontend/app/work_packages/controllers/details-tab-overview-controller.js:
```diff
@@ -77,5 +78,8 @@ module.exports = function(
       });
     });
+    $scope.$on('workPackageUpdatedInEditor', function() {
+      NotificationsService.addSuccess('Successful update');
```
I18n required.

Commit:
inject missing dependency for translations

Signed-off-by: Florian Kraft <f.kraft@finn.de>

ulferts commented:
When deleting an attachment, it gets deleted, yeah. But the corresponding journal entry is not shown in the "LATEST ACTIVITY" section until I reload or add another comment. The journals for deleting themselves are OK by the way.

floriank commented on frontend/app/ui_components/notification-box-directive.js:
```diff
@@ -0,0 +1,74 @@
+
+    scope.$on('upload.error', function() {
+      if (scope.content.type === 'upload') {
+        scope.content.type = 'error';
```
@ulferts: Whenever an upload error happens, the notification will become an error notification

NobodysNightmare replied:
I think the problem @ulferts points out (and I would agree to) is that the general `notification-box-directive` does know and handle words like `upload`. This smells very specific.

floriank replied:
i cannot think of a smarter solution right now, as the `notification-box` handles all the types of notification boxes one may have. 

Strictly speaking I could have multiple components (directives) for each type of notification box - then again, most of the functionality overlaps.

NobodysNightmare commented:
@ulferts regarding:

> When deleting an attachment [...] the corresponding journal entry is not shown [...]

~~I think this is something I need to solve here: https://github.com/opf/openproject/pull/3273~~

Or is the entry being created and just not shown by the frontend?

edit: I just verified and i does not seem to be a backend issue. Still I don't quite understand what makes journals behave as they do ^^

floriank commented:
Reagarding the problem that attachments do not trigger the latest activity to update - this is a problem with the "Latest activities" itself, as this part relies on data being sent in from outside. The component itself is not able to fetch the activities data, unfortunately.

The "data" sent from the outside is generated on reloading the form, which in turn, is currently only triggered by saving a field 9as fields hold all the logic for saving the form itself.

ulferts commented:
@floriank I think one or two tests could help to prevent regressions :).

floriank commented:
argh, daminit and here i was hoping no one would notice.

ulferts commented:
Just a reminder because we talked about it:

![image](https://cloud.githubusercontent.com/assets/617519/8853384/a63c71ca-315a-11e5-9245-cd49997e2ada.png)

This specification in the styleguide deviates from the truth in that nowadays the whole of the notification becomes an error.

ulferts commented:
Very good altogether :bow:

Let's try to get this in as soon as possible and keep questions that have arisen like:
- Additional styling of the files
- Behaviour of the files section

out of this PR for now. They need to be addressed, but they should be addressed in subsequent PRs.

ulferts commented:
One other deviation in the upload notifications from the ones in the styleguide is that the checkmarks are sometimes out of place and have the wrong color (black).

![image](https://cloud.githubusercontent.com/assets/617519/8853643/edef0b26-315b-11e5-9b8f-95e89274f7e9.png)

The file is uploaded and once that finishes, the checkmark is below the progress bar (and black). Once the progress bar vanishes, the checkmark moves up.

Commit:
add tests to verify NotificationsService

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
allow for preprocessing templates in unit tests

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
add basic tests for NotificationsDirective

Signed-off-by: Florian Kraft <f.kraft@finn.de>

NobodysNightmare commented:
Seems like the specs failed due to codecov going nuts: https://gist.github.com/NobodysNightmare/f00f2e919482c7e2ba26

@linki Can we change the test script in a way that the upload of coverage results does not influence the travis outcome? (mentioning you as you introduced codecov in the first place IIRC)

Commit:
add Notification Box directive test

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
add a test for the work package attachments service

Signed-off-by: Florian Kraft <f.kraft@finn.de>

floriank commented:
Styling updates will be done alognside the watchers implementation, as i will create the css component necessary in that context.

Commit:
update tests and fix codestyle problems

Signed-off-by: Florian Kraft <f.kraft@finn.de>

floriank commented:
there is an issue with the mocking of files during testing. I am investigating..

NobodysNightmare commented:
[This comment](https://github.com/opf/openproject/pull/3245#issuecomment-124132412) form @ulferts regarding checkmarks being black and on a new line is still unaddressed.

Do you want to fix that within this PR or separately? If separately: Where is the WP? xD

floriank commented:
Quoting myself:

> Styling updates will be done alognside the watchers implementation, as i will create the css component necessary in that context.
